test-data/unit/check-classes.test

```diff
@@ -3137,3 +3138,51 @@ class M(type):
 class A(metaclass=M): pass
 reveal_type(type(A).x)  # E: Revealed type is 'builtins.int'
 
+[case testMetaclassStrictSupertypeOfTypeWithClassmethods]
+from typing import Type, TypeVar
+TA = TypeVar('TA', bound='A')
+TTA = TypeVar('TTA', bound='Type[A]')
+TM = TypeVar('TM', bound='M')
+
+class M(type):
+    def g1(cls: 'Type[A]') -> None: pass #  E: The erased type of self 'Type[__main__.A]' is not a supertype of its class '__main__.M'
+    def g2(cls: Type[TA]) -> None: pass #  E: The erased type of self 'Type[__main__.A]' is not a supertype of its class '__main__.M'
```

Comment:
Although ``Type[A]`` is lower than ``M``, I think this could be allowed. The special-casing probably should be in ``checkmember.bind_self`` (then you will also not need special-casing ``EnumMeta`` above). My idea is following:
```python
class M(type):
    def g2(cls: Type[TA]) -> TA:
        ...
    def g4(cls: TM) -> None:
        ...
class A(metaclass=M): ...
class A2(A): ...
class B(metaclass=M): ...
A.g2() # OK, inferred type is 'A'
A2.g2() # OK, inferred type is 'A2'
B.g2() # Error, g2 should be only called on subclasses of A
A.g4(), A2.g4(), B.g4() # All are OK
```
The situation with one "privileged" class with a given metaclass sometimes appears in frameworks. This will allow users to give more precise types in such situation.